src/universal/decorators/requireAuthAndRole/requireAuthAndRole.js

```diff
@@ -0,0 +1,43 @@
+import React, {Component, PropTypes} from 'react';
+import {push} from 'react-router-redux';
+import getAuth from 'universal/redux/getAuth';
+import {cashay} from 'cashay';
+import jwtDecode from 'jwt-decode';
+import {error as showError} from 'universal/modules/notifications/ducks/notifications';
+
+const unauthorized = {
+  title: 'Unauthorized',
+  message: 'Hey! You\'re not supposed to be there. Bringing you someplace safe.',
+};
+
+export default role => {
+  return (ComposedComponent) => {
+    return class RequireAuthAndRole extends Component {
+      static propTypes = {
+        lastPathName: PropTypes.string,
+      }
+
+      render() {
+        // no need to check for expired tokens here, right?
+        // We check initial validation in the client.js. Our job should be to keep them logged in.
+        const auth = getAuth();
+        const authObj = auth.authToken && jwtDecode(auth.authToken);
+        if (authObj) {
+          if (role && authObj.rol === role) {
+            // We had a role to check, and role checks out:
+            return <ComposedComponent {...this.props} auth={auth} />;
+          } else if (role === undefined) {
+            // We were looking for any authenticated user only:
+            return <ComposedComponent {...this.props} auth={auth} />;
+          }
+        }
+        // TODO: redirect back from whence they came
```

Comment:
I tried `goBack()`. I tried to make it smart by introspecting on the router store to see if there was any history. If history, `goBack()`, else `push('/')`. It introduced some React spookiness about modifying the store in the middle of a synchronous action. I decided to punt for now...

I think your approach will be much more sound!